### DISCUSSION TITLE
Two questions: How to inventory automatically and how to inventory software

### DISCUSSION THREAD

**Discussion by fthommen** [netbox-community/netbox#7303] [Q&A] [answered]
Dear all,

in my search for an asset inventory software/CMDB I found netbox, which seems to provide almost all features that we are looking for and require. However I am missing two features/mechanisms.  Please don't hesitate to just refer me to the appropriate documentation pages if I have overlooked them somehow:

  * **How can I inventory hosts automatically?** I am looking for some kind of agent/cronjob which regularly gathers the inventory on the host side and pushes it (or the relevant changes) to the CMDB.  How does netbox do that?  A "pull" type of inventory is not usable in our case.  We need some kind "push" mechanism.  Perfect would be to additionally have the possibility to work asynchronously (inventory locally and push later).

  * **How can I inventory software?**  Can this be done at all? Not only we need the hardware information to be included in the CMDB, but also the software packages installed on each host.  I could not find any software inventory being mentioned in the documentation and also didn't find any appropriate plugins.

I am aware that it is possible to write one's own plugins, but I fear that this is nothing we are able to do.  If it were Perl yes, but Python...:-)

Thank you in advance for any hints and pointers

**Comment by candlerb** [comment, marked as answer]
To me, it sounds like Netbox is not going to be the appropriate tool for your needs.  Read the introduction page in the documentation, especially [this](https://netbox.readthedocs.io/en/stable/#serve-as-a-source-of-truth).

As a matter of philosophy, it's expected that you'll use Netbox data to configure your monitoring systems and your devices; not to use your monitoring systems to populate Netbox, and certainly not to use Netbox as a portal onto the "gathered" state of your network, nor to override "desired" state with "actual" state.

You might want to have a look at Netdisco - which incidentally is written in Perl I believe.

**Reply by fthommen** [reply]
Thank you very much.  But I was not mentioning or referring to any monitoring system :-).  I had seen Netbox as an inventory system (for IPAM, issue tracking, monitoring ecc. we use separate and already established tools).  As such the data has to be fed into the inventory somehow, and hopefully not manually (300+ hosts....) :-). But I might have misunderstood the Netbox philosophy.

Netdisco looks too network centric for our use case.  We are searching for a datacenter/rack/server/hardware/software centric tool and from the demo Netbox looked very promising compared to many of the other candidates (see below). The network itself if not of our concern.  We need the inventory and, yes, we plan to finally use this inventory as source of truth for the CMS (Ansible).

I already had a look at GLPI, iTop, Loginventory, Ralph 3, Snipe-IT and RackTables

**Reply by candlerb** [reply]
If you're thinking about a one-time import, it's possible to do this via CSV, or REST API, or Custom Script (the latter uses the Django ORM, which requires writing Python).  The CSV approach can be a bit awkward as you'll need to import Devices, VMs, Interfaces and IP Addresses separately if you want them all.

One of the reasons Netbox doesn't really suit automated updates, apart from the underlying philosophical reasons, is that "Device" and "VirtualMachine" are two disjoint models.  You cannot convert from one to the other, short of deleting and recreating an object of a different type.  So if you auto-detect a system via SNMP or whatever, you're not going to be able to import it unless you know for sure whether it's a Device or a VM, and distinguishing these two may be difficult - unless you got the data from an existing inventory API.

You say you have a separate IPAM that you're keeping, which means maybe all you want from Netbox is Devices.  But it can't model rack layouts unless you tell it which rack units each device sits at, which again can't be auto-detected.

In any case, Netbox isn't a full-blown inventory system.  It doesn't record the history of individual assets - when they were bought, when they were sent for repair or swapped out, nor purchase orders, maintenance contracts etc.

What it models reasonably well is the situation *right now* - that is, where your assets are currently deployed in racks - or are about to be deployed (in "planned" state).  Other useful modelling includes connections (where cables go); and tenancy (ownership/responsibility for each device).

**Reply by fthommen** [reply]
I see.  Thanks a lot.  In this case Netbox - while seeming a well designed tool - in fact doesn't seem to fit our use case.  I was not thinking one-time import but regular, cron-based, push-type updates of the inventory.  We don't do SNMP detection of devices but I admit that I hadn't looked at the history function (yet).  I will continue to look into the other products.

Thanks for your replies.  They were very helpful to assess the (unfortunately non-)usability for our use case ;-)
